### ISSUE TITLE
Kafka consumer groups filling up after upgrade from Sentry 21.2.0 to 21.11.0

### ISSUE THREAD

**Issue by clemens** [getsentry/self-hosted#1204]
### Version

21.11.0

### Steps to Reproduce

- Install Sentry 21.2.0
- Upgrade to Sentry 21.6.3 and then 21.11.0 (as per readme)
- Send some events (e.g. transactions)

### Expected Result

Sentry should work off the events as it used to under version 21.2.0 and not fill up consumer groups.

### Actual Result

Sentry fills up 3 consumer groups (namely `snuba-consumers`, `snuba-events-subscriptions-consumers` and `snub-post-processor`) before dropping (we don't know if the events are processed or simply dropped), see screenshot:

<img width="1358" alt="Screenshot 2021-12-15 at 14 06 14" src="https://user-images.githubusercontent.com/3749/146195101-712897a3-fe17-49ca-b755-af0ae9ef3e99.png">

The period where the consumer groups fill up and drop to almost 0 are irregular

**Comment by clemens** [comment]
Hey guys,

we're having trouble since we upgraded from Sentry `21.2.0` to `21.11.0`. As you can see in the screenshot below, 3 of our Kafa consumer groups (namely `snuba-consumers`, `snuba-events-subscriptions-consumers` and `snub-post-processor`) are filling up quite heavily before suddenly dropping. This has been happening since we updated.

While investigating, we manually ran a command in a snuba container, namely `snuba consumer --storage errors --auto-offset-reset=latest --max-batch-time-ms 750`. As we would have expected, it didn't do anything and just kept waiting. When we changed to command to `snuba consumer --storage events --auto-offset-reset=latest --max-batch-time-ms 750` (so changed the storage from `errors` to `events`), the consumer immediately started working and the lag in the consumer groups went down to almost 0.

We don't really understand why on the one hand it's not working with the original command and on the other hand, we also don't understand why at changing intervals, something actually seems to drop the lag (either by working off messages or dropping them – that we don't know).

We followed the steps from https://develop.sentry.dev/self-hosted/releases/ and also took into account the documented hard stops:

- We first went from `21.2.0` to `21.6.3` and tested the system for general functionality
- Then we went from `21.6.3` to `21.11.0` and tested the system for general functionality

Note that our test for general functionality didn't include checking the Kafka state, so it might well be that the issue already existed after our upgrade to `21.6.3`.

Our only lead with the whole issue is that when doing the upgrade, we found it strange that both versions < `9.1.2` as well as `9.1.2` itself required a hard stop at `21.5.0` while apparently versions > `9.1.2` but < `21.6.3` didn't require that hard stop.

Any help would be much appreciated.

**Comment by chadwhitacre** [comment]
Sorry this fell through the cracks over the holidays @clemens ... reopening. Is this at all related to #1199, I wonder? 🤔

**Comment by ahmad10244** [comment]
> Hey guys,
> 
> we're having trouble since we upgraded from Sentry `21.2.0` to `21.11.0`. As you can see in the screenshot below, 3 of our Kafa consumer groups (namely `snuba-consumers`, `snuba-events-subscriptions-consumers` and `snub-post-processor`) are filling up quite heavily before suddenly dropping. This has been happening since we updated.
> 
> While investigating, we manually ran a command in a snuba container, namely `snuba consumer --storage errors --auto-offset-reset=latest --max-batch-time-ms 750`. As we would have expected, it didn't do anything and just kept waiting. When we changed to command to `snuba consumer --storage events --auto-offset-reset=latest --max-batch-time-ms 750` (so changed the storage from `errors` to `events`), the consumer immediately started working and the lag in the consumer groups went down to almost 0.
> 
> We don't really understand why on the one hand it's not working with the original command and on the other hand, we also don't understand why at changing intervals, something actually seems to drop the lag (either by working off messages or dropping them – that we don't know).
> 
> We followed the steps from https://develop.sentry.dev/self-hosted/releases/ and also took into account the documented hard stops:
> 
> * We first went from `21.2.0` to `21.6.3` and tested the system for general functionality
> * Then we went from `21.6.3` to `21.11.0` and tested the system for general functionality
> 
> Note that our test for general functionality didn't include checking the Kafka state, so it might well be that the issue already existed after our upgrade to `21.6.3`.
> 
> Our only lead with the whole issue is that when doing the upgrade, we found it strange that both versions < `9.1.2` as well as `9.1.2` itself required a hard stop at `21.5.0` while apparently versions > `9.1.2` but < `21.6.3` didn't require that hard stop.
> 
> Any help would be much appreciated.

Same problem for me. After installing v21.12 offset lag start increasing but if any issue occurred on any project offset lag becomes 0 and again start to increase.
I changed storage from `errors` to `events` and offset lag problem solved but new project issues does not capture by sentry.

**Comment by aminvakil** [comment]
> Same problem for me. After installing v21.12 offset lag start increasing but if any issue occurred on any project, offset lag becomes 0 and again start to increase. I changed storage from `errors` to `events` and offset lag problem solved but new project issues does not capture by sentry.

From which version have you upgraded to `21.12.0`? Are you sure this is related to this issue or do you think this should be tracked in another issue?
What do you mean from changing storage from `errors` to `events`?

**Comment by ahmad10244** [comment]
> > Same problem for me. After installing v21.12 offset lag start increasing but if any issue occurred on any project, offset lag becomes 0 and again start to increase. I changed storage from `errors` to `events` and offset lag problem solved but new project issues does not capture by sentry.
> 
> From which version have you upgraded to `21.12.0`? Are you sure this is related to this issue or do you think this should be tracked in another issue? What do you mean from changing storage from `errors` to `events`?

I installed v21.12 from scratch. 
my issue is the same with @clemens  and kafka consumer groups fill up(offset lag increases until `OFFSET_OUT_OF_RANGE` error occurred)
I changed `errors` to `events` on this line:
https://github.com/getsentry/self-hosted/blob/13d14e8db723446696068be733e9fedcc5ec8bc4/docker-compose.yml#L229

**Comment by clemens** [comment]
> Sorry this fell through the cracks over the holidays @clemens ... reopening. Is this at all related to #1199, I wonder? 🤔

From my understanding, they're not related. I've asked a DevOps colleague and here's what he wrote:

> what's discussed in #1199 is not related in my opinion. What is suggested there is to purge kafka because it can't keep up with the events. It's not our case. The VM is not stressed and the kafka container has enough resources. The consumers are working their input in batches, rather than as a continuous stream. This started after the update. If you look [here](https://link-to-internal-logging.dataguard.de), the `snuba-transactions-subscriptions-consumers` and the `transactions_group` seem to do fine in working the events topic, the others don't. Now, if they're meant to work like this, let's cope with that. It's just difficult to monitor: I should trigger an alarm after a queue has more than 100k items for 8 hours. This is 4 orders of magnitude with respect to a normal behaviour of other queues. We'll be able to catch an error after hours.

Here's a screenshot from the internal link he is referencing:

![Screenshot 2022-01-18 at 19 17 55](https://user-images.githubusercontent.com/3749/149995774-84804101-fb1d-460d-b8df-001254d9a2d9.png)

We can't spot any real pattern in the numbers, i.e. it doesn't seem to be that it's e.g. spiking every day at, say, 1am because of some failures in nightly background routines or something like that.

**Comment by taxilian** [comment]
I am also experiencing this issue, but I was seeing it before I upgraded from 21.8.0 to 21.12.0 and upgrading to 21.12.0 didn't help. I'm running in kubernetes and started with the helm chart. I mostly only see the `snuba-events-subscriptions-consumers` group have issues, but have occasionally had the other two run out as well.

CPU usage on all pods seems fine, I've tried the "nuclear option" on kafka and it didn't help.

**Comment by chadwhitacre** [comment]
@clemens @ahmad10244 @taxilian Looks like we've got enough folks here to triangulate. ;) I've dropped a line internally to see if we can get some eyes on this ...

> we found it strange that both versions < 9.1.2 as well as 9.1.2 itself required a hard stop at 21.5.0 while apparently versions > 9.1.2 but < 21.6.3 didn't require that hard stop.

The specifics of the hard stops are due to the migration system available in each version ([South](https://south.readthedocs.io/en/latest/) vs. [Django Migrations](https://docs.djangoproject.com/en/4.0/topics/migrations/)) as well as (iirc) bugs in migrations that were fixed in later versions. It's complicated. I wasn't here for all the history and I don't fully grok it, https://github.com/getsentry/develop/pull/411 is the closest I've come so far. 😅 tl;dr I think that's a red herring wrt this ticket.

**Comment by nikhars** [comment]
Hello everyone

I have a question regarding this behavior. When you are seeing this behavior, are you sending both errors and transactions to Sentry? Or are you sending only one out of errors and transactions? If you are sending only errors or only transactions, can you see if the behavior happens if you send both?

Some context:
Errors consumers and Transactions consumers use a common kafka topic. We recently introduced a performance improvement which filters messages on the consumers based on kafka headers if the message is not intended for the consumer. This causes kafka offsets not to be committed when only one type of message is on the shared kafka topic.

I will tag this as a bug and have it fixed.

**Comment by clemens** [comment]
We‘re using both, errors and transactions in all of our apps. Happy to try and disable transactions if you’re reasonably confident that this might be an issue (we have 10 or so apps using Sentry that would all need adaptations and deployment, which I‘m happy to get done if it might lead to something).

**Comment by nikhars** [comment]
No, please don't disable either errors or transactions. Having both is a good thing. 

And to answer the question about whether data is being dropped or not, Kafka has a default retention period of 7 days. So if the spikes recover within that time frame, data has not been lost.

> We‘re using both, errors and transactions in all of our apps. Happy to try and disable transactions if you’re reasonably confident that this might be an issue (we have 10 or so apps using Sentry that would all need adaptations and deployment, which I‘m happy to get done if it might lead to something).
> […](#)
> On 21.01.2022, at 18:32, Nikhar Saxena ***@***.***> wrote: ﻿ Hello everyone I have a question regarding this behavior. When you are seeing this behavior, are you sending both errors and transactions to Sentry? Or are you sending only one out of errors and transactions? If you are sending only errors or only transactions, can you see if the behavior happens if you send both? Some context: Errors consumers and Transactions consumers use a common kafka topic. We recently introduced a performance improvement which filters messages on the consumers based on kafka headers if the message is not intended for the consumer. This causes kafka offsets not to be committed when only one type of message is on the shared kafka topic. I will tag this as a bug and have it fixed. — Reply to this email directly, view it on GitHub, or unsubscribe. Triage notifications on the go with GitHub Mobile for iOS or Android. You are receiving this because you were mentioned.

**Comment by chadwhitacre** [comment]
👏 @nikhars 

@clemens @ahmad10244 @taxilian Anyone up for trying with the latest nightly to verify the fix?

**Comment by nikhars** [comment]
To provide a little bit more context. The fix removes a kafka header which causes this behavior. So even after you deploy the fix, you might see the same behavior for sometime. This is because there could be messages in the kafka queue which have the old kafka header. Once the kafka queue gets drained, you should start seeing this behavior go away. Hence, try to run Sentry for sometime to verify if the behavior is fixed.

**Comment by ahmad10244** [comment]
> clap @nikhars
> 
> @clemens @ahmad10244 @taxilian Anyone up for trying with the latest nightly to verify the fix?

Install nightly version with this: `SENTRY_IMAGE=getsentry/sentry:nightly ./install.sh` ?
I did that and also reset consumer offsets to latest but the issue did not solve.

**Comment by chadwhitacre** [comment]
Reopening while we discuss ...

**Comment by taxilian** [comment]
Would it be the sentry image that needs to be changed or would it be a snuba container? (I don't understand the infrastructure well). I'm on kubernetes, set up using the helm chart initially though I'm managing it manually now.

**Comment by DaspawnW** [comment]
@taxilian: I think from what I see it's really the getsentry/sentry, as this was the one where the merge request #31292 was made for.

General: I'm also running on Kubernetes and it would be great to know which component needs to be updated to nightly. I would like to prevent to change for all sentry components to a nightly build, this feels not really well.

Thank you in advance,
Björn

**Comment by nikhars** [comment]
Hi folks

The fix was made on the sentry repository. So you will need to update the container running Sentry.

Once the fix is applied in your environment and you still see the behavior can someone run kafkacat/kcat command to fetch the headers and provide that output? Something along these lines
```
kafkacat -b localhost:9092 -X broker.address.family=v4 -t events -C -f '\nHeaders: %h\n'
```

Would also like to know if the option `eventstream:kafka-headers` is enabled or disabled in the environment. The option could be found in the sentry conf file

**Comment by taxilian** [comment]
can't find kafkacat in any of my containers; where would I look for it?

**Comment by ahmad10244** [comment]
> Hi folks
> 
> The fix was made on the sentry repository. So you will need to update the container running Sentry.
> 
> Once the fix is applied in your environment and you still see the behavior can someone run kafkacat/kcat command to fetch the headers and provide that output? Something along these lines
> 
> ```
> kafkacat -b localhost:9092 -X broker.address.family=v4 -t events -C -f '\nHeaders: %h\n'
> ```
> 
> Would also like to know if the option `eventstream:kafka-headers` is enabled or disabled in the environment. The option could be found in the sentry conf file
Hi
I installed nightly version with this: `SENTRY_IMAGE=getsentry/sentry:nightly ./install.sh`
and hear is the output of kafkacat:
`
Headers: Received-Timestamp=1643451454.253542,transaction_forwarder=1,operation=insert,version=2
% Reached end of topic events [0] at offset 1476

Headers: Received-Timestamp=1643451461.261176,transaction_forwarder=1,operation=insert,version=2
% Reached end of topic events [0] at offset 1477

Headers: Received-Timestamp=1643451468.269351,transaction_forwarder=1,operation=insert,version=2
% Reached end of topic events [0] at offset 1478

Headers: Received-Timestamp=1643451475.277626,transaction_forwarder=1,operation=insert,version=2
% Reached end of topic events [0] at offset 1479

Headers: Received-Timestamp=1643451482.288537,transaction_forwarder=1,operation=insert,version=2
% Reached end of topic events [0] at offset 1480
`

**Comment by nikhars** [comment]
> > Hi folks
> > The fix was made on the sentry repository. So you will need to update the container running Sentry.
> > Once the fix is applied in your environment and you still see the behavior can someone run kafkacat/kcat command to fetch the headers and provide that output? Something along these lines
> > ```
> > kafkacat -b localhost:9092 -X broker.address.family=v4 -t events -C -f '\nHeaders: %h\n'
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Would also like to know if the option `eventstream:kafka-headers` is enabled or disabled in the environment. The option could be found in the sentry conf file
> 
> Hi I installed nightly version with this command: `SENTRY_IMAGE=getsentry/sentry:nightly ./install.sh` and here is the output of kafkacat:
> 
> ```
> Headers: Received-Timestamp=1643451454.253542,transaction_forwarder=1,operation=insert,version=2
> % Reached end of topic events [0] at offset 1476
> 
> Headers: Received-Timestamp=1643451461.261176,transaction_forwarder=1,operation=insert,version=2
> % Reached end of topic events [0] at offset 1477
> 
> Headers: Received-Timestamp=1643451468.269351,transaction_forwarder=1,operation=insert,version=2
> % Reached end of topic events [0] at offset 1478
> 
> Headers: Received-Timestamp=1643451475.277626,transaction_forwarder=1,operation=insert,version=2
> % Reached end of topic events [0] at offset 1479
> 
> Headers: Received-Timestamp=1643451482.288537,transaction_forwarder=1,operation=insert,version=2
> % Reached end of topic events [0] at offset 1480
> ```

With the change in PR https://github.com/getsentry/sentry/pull/31292, the header `transaction_forwarder` should no longer be present on the messages. Do you see the `transaction_forwarder` header in the latest message offsets as well? Can you check if you have the following option set in your configurations `eventstream:kafka-headers`?

In my environment with the fix in place the output looks like this.
`Headers: Received-Timestamp=1643756230.091956,operation=insert,version=2

Headers: Received-Timestamp=1643756290.087899,operation=insert,version=2

Headers: Received-Timestamp=1643756350.085775,operation=insert,version=2

Headers: Received-Timestamp=1643756400.08789,operation=insert,version=2

Headers: Received-Timestamp=1643756410.098044,operation=insert,version=2

Headers: Received-Timestamp=1643756470.088023,operation=insert,version=2
% Reached end of topic events [0] at offset 22581`

**Comment by taxilian** [comment]
To provide an additional datapoint, I just double checked that every sentry container I have running is using getsentry/sentry:latest and that it pulled the latest before starting it...

```
Headers: Received-Timestamp=1643757214.514335,transaction_forwarder=1,operation=insert,version=2

Headers: Received-Timestamp=1643757214.17919,transaction_forwarder=1,operation=insert,version=2

Headers: Received-Timestamp=1643757214.625791,transaction_forwarder=1,operation=insert,version=2

Headers: Received-Timestamp=1643757214.701462,transaction_forwarder=1,operation=insert,version=2

Headers: Received-Timestamp=1643757216.099573,transaction_forwarder=1,operation=insert,version=2

Headers: Received-Timestamp=1643757217.151305,transaction_forwarder=1,operation=insert,version=2
```

I dug through my configuration and couldn't find anything mentioning `eventstream:kafka-headers` one way or the other, but I may just not know where to look.

I have confirmed however that the backend.py file in the container has the fix in the referenced PR.

**Comment by nikhars** [comment]
> To provide an additional datapoint, I just double checked that every sentry container I have running is using getsentry/sentry:latest and that it pulled the latest before starting it...
> 
> ```
> Headers: Received-Timestamp=1643757214.514335,transaction_forwarder=1,operation=insert,version=2
> 
> Headers: Received-Timestamp=1643757214.17919,transaction_forwarder=1,operation=insert,version=2
> 
> Headers: Received-Timestamp=1643757214.625791,transaction_forwarder=1,operation=insert,version=2
> 
> Headers: Received-Timestamp=1643757214.701462,transaction_forwarder=1,operation=insert,version=2
> 
> Headers: Received-Timestamp=1643757216.099573,transaction_forwarder=1,operation=insert,version=2
> 
> Headers: Received-Timestamp=1643757217.151305,transaction_forwarder=1,operation=insert,version=2
> ```
> 
> I dug through my configuration and couldn't find anything mentioning `eventstream:kafka-headers` one way or the other, but I may just not know where to look.
> 
> I have confirmed however that the backend.py file in the container has the fix in the referenced PR.

This has me stumped. I would recommend the following. Wait for the next release of sentry/self-hosted to happen. When it is released, please try out with that.

**Comment by taxilian** [comment]
> This has me stumped. I would recommend the following. Wait for the next release of sentry/self-hosted to happen. When it is released, please try out with that.

Any suggestions for a dirty hack to keep things at least running? I end up resetting the offset about once a day when my containers go into crash loops.

    kubectl -n sentry exec -it statefulset/sentry-kafka -- kafka-consumer-groups.sh --bootstrap-server sentry-kafka:9092 --group snuba-events-subscriptions-consumers --reset-offsets --to-earliest --all-topics --execute

**Comment by ahmad10244** [comment]
the issue seems to be solved in version 22.2.

**Comment by chadwhitacre** [comment]
Well that's good news. Thanks @ahmad10244! :)

Let's leave this open for a few more days to give others time to report back before we declare victory.

**Comment by taxilian** [comment]
I didn't notice 22.2 was out, so let me give that a try and we'll see what it looks like =]

**Comment by clemens** [comment]
So we're on 22.2.0 and the consumer groups don't seem to fill up anymore, however, there now seem to be 2 containers consuming excessive amounts of memory:

![Screenshot 2022-03-16 at 14 47 20](https://user-images.githubusercontent.com/3749/158604251-f935830e-6f98-443e-a445-436649bd630a.png)

The upper line that constantly keeps increasing is `snuba-subscription-consumer-transactions_production` and the lower line that constantly increases (except that one drop at March 11) is `snuba-subscription-consumer-events_production`.

I'm unsure if these are related to the issue at hand => I'm happy to create a new issue if warranted.

**Comment by nikhars** [comment]
> So we're on 22.2.0 and the consumer groups don't seem to fill up anymore, however, there now seem to be 2 containers consuming excessive amounts of memory:
> 
> ![Screenshot 2022-03-16 at 14 47 20](https://user-images.githubusercontent.com/3749/158604251-f935830e-6f98-443e-a445-436649bd630a.png)
> 
> The upper line that constantly keeps increasing is `snuba-subscription-consumer-transactions_production` and the lower line that constantly increases (except that one drop at March 11) is `snuba-subscription-consumer-events_production`.
> 
> I'm unsure if these are related to the issue at hand => I'm happy to create a new issue if warranted.

The increase in memory has nothing to do with this particular issue. Please report a new issue.

**Comment by chadwhitacre** [comment]
> Let's leave this open for a few more days to give others time to report back before we declare victory.

Ahem. I guess let's declare victory here, then. 😁  🎉

**Comment by ajacques** [comment]
I also confirm that after upgrading from 22.1.0 to 22.2.0, I no-longer see the `snuba-events-subscriptions-consumers` consumer group lagging.

**Comment by taxilian** [comment]
yeah, it's working for me as well. thanks!
